Deep Support in Resample

danieldresser-ie opened the pull request:
It was a bit of a scramble to try and get all the various options explored for this cleaned up into something reviewable.

Hopefully this makes a fair bit of sense, though some of the comments were written pretty hastily.

The main known thing that needs adding is the mode selector to switch between proper filtering or forcing the "nearest" filter for deeps ( and possibly also the option of throwing an exception if you try to deep filter something without choosing "nearest" ). That should be pretty straightforward to add, and shouldn't affect the review much.

danieldresser-ie commented:
Cleaned up the missing test files, and added the "deepMode" plug.
I also split the Resample commit, so there's a first commit that adds "nearest" mode ( I didn't want to put too much work into splitting, so it also contains some general cleanup - mostly making sure that plugs that are supposed to be global, like boundingMode, are evaluated without the tileOrigin in the context ).

The remaining test failure is a failed assertion in DeepState. While I have made a few modifications to DeepState, I suspect this failure is likely not new - ResampleTest.testDeepWithExceptionalValues hits a lot of corner cases in DeepState that were previously untested. I had already noted a test that I skipped because DeepState output completely unreasonable values ... the failed assertion is quite possibly related, and a good excuse to clean up DeepState.

If you want to visually inspect what's going on, there's a WIP version of the deep pixel inspector here: https://github.com/danieldresser-ie/gaffer/tree/deepResizeWithViz
It's full of constuction sawdust since I haven't even removed all the code from the earlier version which predated official image comparison and worked quite differently, but it's already useful, and hopefully won't take too long to clean up.

johnhaddon reviewed (commented):
Thanks Daniel!

I've run out of steam a bit on this review, but I think the reality is that even with a full head, I'm not really equipped to review the finer details. So it's mostly a smattering of suggestions for tweaks to the presentation in terms of API and presentation in the nodes. The most useful contribution is probably the snippet with the Resize node that triggers the "this would be a hang if it wasn't an exception" thing.

Another couple of comments that I didn't comment on inline :

- Cancellation can be excessively slow, with downsizing to a single tile showing the problem particularly badly. Worth tuning the frequency of checks to that case I think, as otherwise the UI can freeze.
- The fact that `Default` filter mode chooses a downsizing filter with negative lobes is vexing, since it leads straight to an error as soon as you turn on filtering. "Default" here meant "makes a sensible decision for you based on the operation", so I think we should interpret that to mean "chooses different filters suitable for deep" too.

Cheers...
John

johnhaddon commented:
Oh, also spotted we have this failure in the debug CI :

```
testDeepWithExceptionalValues (GafferImageTest.ResampleTest.ResampleTest) ... gaffer: src/GafferImage/DeepState.cpp:326: IECore::FloatVectorDataPtr {anonymous}::alphaToLinearWeights(std::vector<float>&, const std::vector<int>&, const std::vector<int>&, const std::vector<float>&, const std::vector<int>&, bool): Assertion `contributionEnd != contributionStart' failed.
gaffer: src/GafferImage/DeepState.cpp:326: IECore::FloatVectorDataPtr {anonymous}::alphaToLinearWeights(std::vector<float>&, const std::vector<int>&, const std::vector<int>&, const std::vector<float>&, const std::vector<int>&, bool): Assertion `contributionEnd != contributionStart' failed.
```

danieldresser-ie commented on src/GafferImage/Resample.cpp:
```diff
@@ -518,18 +548,28 @@ void Resample::hashChannelData( const GafferImage::ImagePlug *parent, const Gaff
+	const V2i tileOrigin = context->get<V2i>( ImagePlug::tileOriginContextName );
+	const std::string &channelName = context->get<std::string>( ImagePlug::channelNameContextName );
+	Passes passes = requiredPasses( this, parent, filter, ratio );
+	Box2i ir = inputRegion( tileOrigin, passes, ratio, offset, filter, inputFilterScale );
```
This is where it's a little bit odd how I ended up splitting these commits - I would agree with you if this was the final form of the code, but once I add the deep stuff, this stuff is needed for the deep path before we decide to construct a 2D sampler.

danieldresser-ie commented on include/GafferImage/DeepTileAccessor.h:
```diff
@@ -0,0 +1,109 @@
+		/// that this point is contained within the sample window passed to the
+		/// constructor. If this was constructed with an empty channel name, data will
+		/// be set to nullptr
+		void sample( int x, int y, const float* &data, unsigned int &count );
```
Sure, I'm happy to try out boost span and a separate sampleCount( x, y ). You're happy with the overall interface, with 'pass in channelName = "" if you just want counts"? I guess I can make sample() throw if you constructed it with an empty channel?

johnhaddon replied:
> I guess I can make sample() throw if you constructed it with an empty channel?

Sounds good. Or it could just be UB, like when you pass a position which is outside the sample window.

danieldresser-ie commented on include/GafferImage/DeepTileAccessor.h:
```diff
@@ -0,0 +1,109 @@
+
+/// Utility class for sampling pixel values from a deep image.
+/// The interface mostly matches Sampler for flat images.
+class GAFFERIMAGE_API DeepTileAccessor
```
Sure, I'm fine with DeepPixelAccessor. Does that mean that Sampler might become "PixelAccessor" in the future?

johnhaddon replied:
> Does that mean that Sampler might become "PixelAccessor" in the future?

Yep. I think cleaning up the Sampler/ImageSampler confusion would be worthwhile. Maybe we do it once this is merged?

danieldresser-ie commented on python/GafferImageUI/PremultiplyUI.py:
```diff
@@ -65,6 +65,20 @@
 
 		],
 
+		"useDeepVisibility" : [
+
+			"description",
+			"""
+			When processing a deep image, you may use this to multiply by
+			the visibility of the current sample, taking into account the
+			alpha of all previous samples. This is a pretty special case,
+			it's mostly useful for converting deep images to incandescence,
+			by multiplying RGB by visibility, and then wiping out the 'A'
+			channel.
+			""",
+
+		],
```
To be honest, this is really quite specific, and I'm not sure how realistic a use case for it is ... I just wanted to stick the code I needed to implement the test somewhere without spending time making a new node. If you want me to make a DeepIncandescence, I can do that. Another alternative would be just putting it in GafferImageTest ... though if I'm going to make a DeepIncandesence node, I guess I might as well expose it - it does do a good job of a specific task that someone could want to do ( it's just not likely to be at all common ).

danieldresser-ie commented on src/GafferImage/Resample.cpp:
```diff
@@ -318,6 +345,626 @@ Box2f transform( const Box2f &b, const M33f &m )
+			// triggered by a weird floating point value or something, this could lead to a hang.
+			// We definitely don't want to hang - if we complete an iteration without progress, throw
+			// an exception instead.
+			throw IECore::Exception(
+				"Internal failure in Resample while processing deep image. This should "
+				"not be possible. We would appreciate if you can isolate a crop of the image which triggers "
+				"the issue and submit it to the Gaffer dev team."
+			);
```
This turned out to be quite a simple thing ... I'd copied the code for filterWeights from filterWeights1D, and expected it to match filterWeights2D, but there's what is probably a typo in there - a constant that should be "0.5f" is just "0.5", causing the calculation to happen in double precision, and get a non-matching result do to precision.

The other thing that came up when playing with this is that performance of resizing to a very small result is incredibly bad. I probably shouldn't worry about this for now, because it's not specific to how we handle deep: Resample matches a common convention where we don't worry about the dataWindow while processing tiles. This means that due to the tile size, resizing a 1000 pixel image down to 1 pixel is treated identically to resizing a 128000 pixel image down to 128 pixels. This is actually kinda less of a problem for deep images with boundingMode set to Black, because the deep processing is so much slower on pixels that aren't empty ( whereas using a non-seperable filter to resize a 2D image down to one pixel would be extremely awful for no reason ), but if boundingMode is set to Clamp ( as it is by resize ), then we invent piles of extra pixels to filter together to fill in all the parts of the tile that will never be read.

Probably not high urgency, since the immediate use we expect for deep is upsampling. Eventually, we should probably try to fix it for both deep and flat ( while being careful with how perf sensitive the flat case is ).

johnhaddon replied:
> The other thing that came up when playing with this is that performance of resizing to a very small result is incredibly bad.

The other part of this is that when the output image is only one tile, we lose out on multithreading. So for deep at least, it seems like part of the solution might be to do some threading internal to a tile? But I agree this isn't high urgency at present - definitely not something for this PR.

danieldresser-ie commented on include/GafferImage/Resample.h:
```diff
@@ -72,6 +72,13 @@ class GAFFERIMAGE_API Resample : public FlatImageProcessor
 
+		enum class DeepMode
+		{
+			Accurate,
+			ForceNearest,
+			ErrorIfSlow
```
I'm willing to abandon the 3rd mode if it makes things too complicated ... I'd already said it's of limited utility if it's not the default, and the main point would be to point users to more efficient options ... whether we even offer more efficient options is another conversation.

I feel pretty strongly about keeping Nearest as a filter though. In many uses of deep images, you have flat aovs or beauty images that you want to combine with the Deep - you may need to resize those as well, and in order to keep them matching, you need resize them with a matching filter.

It also feels like a pretty important diagnostic tool when figuring out deep stuff - if someone is unhappy with the quality of there results, it feels quite useful to say "Well, if you do it on flat images using nearest, you should get the same result", which helps narrow down whether something unique to deep has gone wrong.

( Having a nearest mode is also very useful as a performance target, flat image resizing isn't that expensive, but our box filter is measurably slower, and I think we should actually be able to reduce that gap - but I know this probably isn't persuasive so I'm trying not to focus on it. )

johnhaddon replied:
> I feel pretty strongly about keeping Nearest as a filter though.

OK, I buy the diagnostic argument.

> I'm willing to abandon the 3rd mode if it makes things too complicated

That's a long way from where we started with this, but I think it'd sit OK with me. Are we now just talking about a `filterDeep` checkbox then?

danieldresser-ie commented on src/GafferImage/Resample.cpp:
```diff
@@ -180,13 +186,28 @@ Box2i inputRegion( const V2i &tileOrigin, unsigned passes, const V2f &ratio, con
+	if( deep && deepMode == Resample::DeepMode::ErrorIfSlow && name != g_nearestString )
+	{
+		throw IECore::Exception(
+			"Accurately filtering deep images is very slow and produces results that are impractically large "
+			"to save to disk. We recommend switching to the filter \"nearest\", using a more efficient "
+			"approach we haven't implemented yet, or if you know what you're doing, and want a fully "
+			"filtered result, set \"deepMode\" to \"Accurate\"."
```
Last we spoke, I thought you had floated the idea of an "all-in-one-merge/flatten" as a separate node network we would include? That would open the door to doing some optimization. Even without any external visibility on the code inside Resample, we could identify whole tiles where only one of the images has content, and skip the deep resize. It also would take very little code to add a mask input to Resample and do that bypass per pixel. There are pretty common production cases where this would be a 5X speedup.

johnhaddon replied:
> Last we spoke, I thought you had floated the idea of an "all-in-one-merge/flatten" as a separate node network we would include? 

That does ring a bell, and seems reasonable as long as it doesn't complicate the internals of some already quite complicated nodes. I still don't think it makes much sense to talk about this in the docs though.

danieldresser-ie commented:
> Oh, also spotted we have this failure in the debug CI :

Yeah, that's the probably pre-existing buy in DeepState I mentioned in my comment when I updated.

> The fact that Default filter mode chooses a downsizing filter with negative lobes is vexing, since it leads straight to an error as soon as you turn on filtering. "Default" here meant "makes a sensible decision for you based on the operation", so I think we should interpret that to mean "chooses different filters suitable for deep" too.

The part we can definitely agree on is that this is incredibly vexing. I'm really not sure what the right solution is those. If you have a flat beauty, then you need to Resize it with a filter that matches your deep alpha. It seem like a pretty nasty trap if you resize both  of them with Default, and then get silhouettes that don't match.

danieldresser-ie commented:
Thinking about how vexing it is to not be able to match a default downsize ... now that the code is structured this way, I think it would actually be easier to do something reasonable with negative filter lobes ( a result that matches anywhere the alpha is in range, and does something reasonable outside of that ) - I could probably get it working in a day.

I know you don't want to spend any more time on this, but it might be more useful than spending time trying to come up with an approximation without negative filter lobes.

danieldresser-ie commented:
Added a bunch of little fix commits to address feedback ... there's a lot of little things, but I think they're all fairly clear what they're doing, so it might be easier to review this way. Easy to squash if you prefer.

Left to do: useDeepVisibility - you didn't actually confirm that you think it's worth splitting this out to be a ToIncandescent node, and figuring out what to do with negative lobes.

danieldresser-ie commented:
Add a few more commits, to handle negative lobes, and useDeepVisibility - Murray voted for just moving it to an advanced tab.

johnhaddon commented:
As discussed on Thursday, I've put together a small commit that adds a warning in the UI when `filterDeep` is on :
https://github.com/GafferHQ/gaffer/commit/91c1fe0169857e55e6accf02d743a3ff1bf227dd. If you could cherry pick that over that'd be great.

johnhaddon commented on python/GafferImageTest/ResampleTest.py:
```diff
@@ -583,7 +599,26 @@ def testBruteForceDeep( self ):
+						# The negative lobes also result in alphas over 1, which makes the numerical precision
+						# for the RGB channels much trickier - it's possible to reach an alpha of 1 before all
+						# contributions are accounted for, which means that all remaining contributions must
+						# be crammed into the last segment in order preserve the final result ... if the last
+						# segment has low visibility, we end up having to put extremely high RGB values in
+						# the last segment, resulting in poor precision. We could be a bit more specific in
+						# this test about where this poor precision happens, but currently we just use a large
+						# tolerance for everything.
```
This makes it sound like although we might be arriving at the expected flattened result, we could be generating deeps that don't make a great deal of sense in themselves, and might not combine as expected with other deeps. The next segment that we push things into could be a long way away from the original, right? Would that be fair to say? If so, I'm still wondering if it really makes sense to have a negative-lobed downsize be the default, and if allowing negative lobes at all might be a bit of a trap...

danieldresser-ie replied:
"The next segment that we push things into could be a long way away from the original, right?" - only if there are no samples in between that increase the alpha back to an increasing value. Most reasonable images don't have a lot of contribution from segments with low alpha, so it isn't a substantial error.

It seemed convincing when I described this and showed how it worked pretty well on a reasonable test image this morning, so I probably should have just left it there, but it did feel like I should maybe quantify the worst case. Which ended up with me slipping in this little fix: 9a69817, and putting together this test:

<details>
<summary>Test Script</summary>
```
import Gaffer
import GafferImage
import GafferOSL
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 4, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Constant"] = GafferImage.Constant( "Constant" )
parent.addChild( __children["Constant"] )
__children["Constant"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["OSLImage"] = GafferOSL.OSLImage( "OSLImage" )
parent.addChild( __children["OSLImage"] )
__children["OSLImage"]["channels"].addChild( Gaffer.NameValuePlug( "A", Gaffer.FloatPlug( "value", defaultValue = 1.0, ), True, "channel", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["OSLImage"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["OSLCode"] = GafferOSL.OSLCode( "OSLCode" )
parent.addChild( __children["OSLCode"] )
__children["OSLCode"]["out"].addChild( Gaffer.FloatPlug( "output1", direction = Gaffer.Plug.Direction.Out, defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["OSLCode"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["DeepMerge1"] = GafferImage.DeepMerge( "DeepMerge1" )
parent.addChild( __children["DeepMerge1"] )
__children["DeepMerge1"]["in"].addChild( GafferImage.ImagePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["DeepMerge1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["FlatToDeep"] = GafferImage.FlatToDeep( "FlatToDeep" )
parent.addChild( __children["FlatToDeep"] )
__children["FlatToDeep"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Premultiply1"] = GafferImage.Premultiply( "Premultiply1" )
parent.addChild( __children["Premultiply1"] )
__children["Premultiply1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["FlatToDeep1"] = GafferImage.FlatToDeep( "FlatToDeep1" )
parent.addChild( __children["FlatToDeep1"] )
__children["FlatToDeep1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SliceAfter"] = GafferImage.DeepSlice( "SliceAfter" )
parent.addChild( __children["SliceAfter"] )
__children["SliceAfter"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["DeepResize"] = GafferImage.Resize( "DeepResize" )
parent.addChild( __children["DeepResize"] )
__children["DeepResize"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["FlatResize"] = GafferImage.Resize( "FlatResize" )
parent.addChild( __children["FlatResize"] )
__children["FlatResize"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Dot"] = Gaffer.Dot( "Dot" )
parent.addChild( __children["Dot"] )
__children["Dot"].setup( GafferImage.ImagePlug( "in", ) )
__children["Dot"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ReasonableDeepImage"] = GafferImage.ImageReader( "ReasonableDeepImage" )
parent.addChild( __children["ReasonableDeepImage"] )
__children["ReasonableDeepImage"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SliceBefore"] = GafferImage.DeepSlice( "SliceBefore" )
parent.addChild( __children["SliceBefore"] )
__children["SliceBefore"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Constant"]["format"].setValue( GafferImage.Format( 200, 200, 1.000 ) )
Gaffer.Metadata.registerValue( __children["Constant"]["format"], 'formatPlugValueWidget:mode', 'custom' )
__children["Constant"]["color"].setValue( imath.Color4f( 0, 1, 1, 0 ) )
__children["Constant"]["__uiPosition"].setValue( imath.V2f( -228.642715, 38.4070778 ) )
__children["OSLImage"]["defaultFormat"].setValue( GafferImage.Format( 200, 200, 1.000 ) )
Gaffer.Metadata.registerValue( __children["OSLImage"]["defaultFormat"], 'formatPlugValueWidget:mode', 'custom' )
__children["OSLImage"]["channels"]["channel"]["value"].setInput( __children["OSLCode"]["out"]["output1"] )
__children["OSLImage"]["__uiPosition"].setValue( imath.V2f( -204.351334, 36.8719368 ) )
__children["OSLCode"]["code"].setValue( 'point q = point( u - 0.5, v - 0.5, 0  );\noutput1 = 1 - smoothstep( 0.48, 0.5, length( q ) );' )
__children["OSLCode"]["__uiPosition"].setValue( imath.V2f( -216.451874, 41.0258522 ) )
__children["DeepMerge1"]["in"][0].setInput( __children["FlatToDeep"]["out"] )
__children["DeepMerge1"]["in"][1].setInput( __children["FlatToDeep1"]["out"] )
__children["DeepMerge1"]["__uiPosition"].setValue( imath.V2f( -215.007034, 17.3797455 ) )
__children["FlatToDeep"]["in"].setInput( __children["Constant"]["out"] )
__children["FlatToDeep"]["zBackMode"].setValue( 1 )
__children["FlatToDeep"]["thickness"].setValue( 1.0 )
__children["FlatToDeep"]["__uiPosition"].setValue( imath.V2f( -228.642715, 30.4069881 ) )
__children["Premultiply1"]["in"].setInput( __children["OSLImage"]["out"] )
__children["Premultiply1"]["__uiPosition"].setValue( imath.V2f( -204.351334, 31.2078724 ) )
__children["FlatToDeep1"]["in"].setInput( __children["Premultiply1"]["out"] )
__children["FlatToDeep1"]["depth"].setValue( 0.800000011920929 )
__children["FlatToDeep1"]["__uiPosition"].setValue( imath.V2f( -204.351334, 25.543808 ) )
__children["SliceAfter"]["in"].setInput( __children["DeepResize"]["out"] )
__children["SliceAfter"]["farClip"]["enabled"].setValue( True )
__children["SliceAfter"]["farClip"]["value"].setValue( 6.956500053405762 )
__children["SliceAfter"]["__uiPosition"].setValue( imath.V2f( -215.007034, -10.7726736 ) )
__children["DeepResize"]["in"].setInput( __children["Dot"]["out"] )
__children["DeepResize"]["format"].setValue( GafferImage.Format( 100, 150, 1.000 ) )
Gaffer.Metadata.registerValue( __children["DeepResize"]["format"], 'formatPlugValueWidget:mode', 'custom' )
__children["DeepResize"]["filterDeep"].setValue( True )
__children["DeepResize"]["__uiPosition"].setValue( imath.V2f( -215.007034, 2.63339186 ) )
__children["FlatResize"]["in"].setInput( __children["SliceBefore"]["out"] )
__children["FlatResize"]["format"].setInput( __children["DeepResize"]["format"] )
Gaffer.Metadata.registerValue( __children["FlatResize"]["format"], 'formatPlugValueWidget:mode', 'custom' )
__children["FlatResize"]["filterDeep"].setValue( True )
__children["FlatResize"]["__uiPosition"].setValue( imath.V2f( -198.92955, -10.9095926 ) )
__children["Dot"]["in"].setInput( __children["DeepMerge1"]["out"] )
__children["Dot"]["__uiPosition"].setValue( imath.V2f( -209.648041, 9.96542358 ) )
__children["ReasonableDeepImage"]["fileName"].setValue( '${GAFFER_ROOT}/python/GafferImageTest/images/representativeDeepImage.exr' )
__children["ReasonableDeepImage"]["__uiPosition"].setValue( imath.V2f( -191.552353, 16.9537373 ) )
__children["SliceBefore"]["in"].setInput( __children["Dot"]["out"] )
__children["SliceBefore"]["farClip"].setInput( __children["SliceAfter"]["farClip"] )
__children["SliceBefore"]["flatten"].setValue( True )
__children["SliceBefore"]["__uiPosition"].setValue( imath.V2f( -198.92955, 2.63339233 ) )


del __children
```
</details>

This is intentionally constructed as an absolute worst case - the segments with color are uniform incandescent fog with no alpha, whereas the segments that drive the alpha have no color, and have a hard edge, in behind the fog. This worst case is indeed pretty ugly ... and looking at it now, I probably could come up with a way to give better results in this case of mixed incandescent/opaque with negative lobes. But if you swap to the image reader in the test script, you can also see how it gives reasonable results for a more reasonable image. Unless you want to roll back to just throwing an exception by default for downsampling, so that people know they need to change their flat filtering as well to match, I still think it's the right move to deploy like this - it gives good results in the vast majority of cases, and if people do end up running into issues specifically with downsampling mixtures of incandescent fog and opaque objects, there is probably room to improve this in the future.

johnhaddon commented:
> Add a few more commits, to handle negative lobes, and useDeepVisibility - Murray voted for just moving it to an advanced tab.

Thanks Daniel. I've been through and resolved all my previous comments, bar one in Premultiply.cpp that seems unaddressed. Also added a query inline about the negative lobe stuff...

danieldresser-ie commented:
Squashed all prior fixes, added 337e0c9 and b2d695c

I've probably gone too verbose with the requested message about negative lobes, never sure how much detail to include here ... I guess the important thing is that if you are getting weird results with an intense mixture of incandescent and non-incandescent samples, then try Blackman-Harris ... though I currently in the tooltip I haven't described at all the issue with "curve shape error" that I discuss in the tests - the same mixture of incandescent and non-incandescent is likely to introduce some of that error, even without negative lobes.

johnhaddon commented:
Thanks Daniel! I've squashed the latest fixups in and added changelog entries. Merging...
